src/node/txorphanage.h

```diff
@@ -0,0 +1,154 @@
+ * global limits are reached, we continuously evict the oldest announcement (sorting non-reconsiderable orphans before
+ * reconsiderable ones) from the most resource-intensive peer until we are back within limits.
+ * - Peers can exceed their individual limits (e.g. because they are very useful transaction relay peers) as long as the
+ *   global limits are not exceeded.
+ * - As long as the orphan has 1 announcer, it remains in the orphanage.
+ * - No peer can trigger the eviction of another peer's orphans.
+ * - Peers' orphans are effectively protected from eviction as long as they don't exceed their limits.
+ * Not thread-safe. Requires external synchronization.
+ */
+class TxOrphanage {
+public:
+    using Usage = int64_t;
+    using Count = unsigned int;
+
+    /** Allows providing orphan information externally */
+    struct OrphanTxBase {
+        CTransactionRef tx;
+        /** Peers added with AddTx or AddAnnouncer. */
+        std::set<NodeId> announcers;
+
+        // Constructor with moved announcers
+        OrphanTxBase(CTransactionRef tx, std::set<NodeId>&& announcers) :
+            tx(std::move(tx)),
+            announcers(std::move(announcers))
+        {}
+    };
+
+    virtual ~TxOrphanage() = default;
+
+    /** Add a new orphan transaction */
+    virtual bool AddTx(const CTransactionRef& tx, NodeId peer) = 0;
+
+    /** Add an additional announcer to an orphan if it exists. Otherwise, do nothing. */
+    virtual bool AddAnnouncer(const Wtxid& wtxid, NodeId peer) = 0;
+
+    /** Get a transaction by its witness txid */
+    virtual CTransactionRef GetTx(const Wtxid& wtxid) const = 0;
+
+    /** Check if we already have an orphan transaction (by wtxid only) */
+    virtual bool HaveTx(const Wtxid& wtxid) const = 0;
+
+    /** Check if a {tx, peer} exists in the orphanage.*/
+    virtual bool HaveTxFromPeer(const Wtxid& wtxid, NodeId peer) const = 0;
+
+    /** Extract a transaction from a peer's work set, and flip it back to non-reconsiderable.
+     *  Returns nullptr if there are no transactions to work on.
+     *  Otherwise returns the transaction reference, and removes
+     *  it from the work set.
+     */
+    virtual CTransactionRef GetTxToReconsider(NodeId peer) = 0;
+
+    /** Erase an orphan by wtxid, including all announcements if there are multiple.
+     * Returns true if an orphan was erased, false if no tx with this wtxid exists. */
+    virtual bool EraseTx(const Wtxid& wtxid) = 0;
+
+    /** Maybe erase all orphans announced by a peer (eg, after that peer disconnects). If an orphan
+     * has been announced by another peer, don't erase, just remove this peer from the list of announcers. */
+    virtual void EraseForPeer(NodeId peer) = 0;
+
+    /** Erase all orphans included in or invalidated by a new block */
+    virtual void EraseForBlock(const CBlock& block) = 0;
+
+    /** Limit the orphanage to DEFAULT_MAX_ORPHAN_TRANSACTIONS. */
```

Comment:
This comment refers to DEFAULT_MAX_ORPHAN_TRANSACTIONS which is not used anymore AFAIU